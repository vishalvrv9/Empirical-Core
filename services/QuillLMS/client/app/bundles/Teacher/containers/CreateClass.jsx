import React from 'react';
import $ from 'jquery';
import DropdownButton from 'react-bootstrap/lib/DropdownButton';
import MenuItem from 'react-bootstrap/lib/MenuItem';
import NumberSuffix from '../components/modules/numberSuffixBuilder.js';
import ButtonLoadingIndicator from '../components/shared/button_loading_indicator.jsx';
import GoogleClassroomModal from '../components/dashboard/google_classroom_modal';

export default React.createClass({

  propTypes: {
    hasActivities: React.PropTypes.bool,
  },

  componentDidMount() {
    this.getClassCode();
  },

  getInitialState() {
    return {
      classroom: {
        name: '',
        grade: null,
        code: null,
      },
      loading: false,
      errors: '',
      showModal: false,
    };
  },

  getClassCode() {
    $.ajax({ url: '/teachers/classrooms/regenerate_code', success: this.setClassCode, });
  },

  setClassCode(data) {
    const newClass = Object.assign({}, this.state.classroom);
    newClass.code = data.code;
    this.setState({ classroom: newClass, loading: false, });
  },

  grades() {
    // Note to maintainers: if you update the grade options here, please also
    // ensure to do so in the following locations:
    //   - /app/views/teachers/students/index.html.erb
    //   - /client/app/bundles/Teacher/components/google_classroom/google_classroom_sync/GoogleClassroomsList.jsx
    const grades = [];
    for (let grade = 1; grade <= 12; grade++) {
      grades.push(
        <MenuItem key={grade} eventKey={grade}>{NumberSuffix(grade)}</MenuItem>
            );
    }
    grades.push(<MenuItem key={'University'} eventKey={'University'}>University</MenuItem>);
    grades.push(<MenuItem key={'Other'} eventKey={'Other'}>Other</MenuItem>);
    return grades;
  },

  handleChange(e) {
    const newClass = Object.assign({}, this.state.classroom);
    newClass.name = e.currentTarget.value;
    this.setState({ classroom: newClass, });
  },

  handleSelect(e) {
    const newClass = Object.assign({}, this.state.classroom);
    newClass.grade = e;
    this.setState({ classroom: newClass, });
  },

  clickCreateAClass() {
    const classroom = this.state.classroom;
    if (classroom.name && classroom.grade) {
      this.submitClassroom();
    } else {
      const missingValue = classroom.name ? 'grade' : 'name';
      this.setState({ errors: `You must choose your classroom's ${missingValue} before continuing.`, });
    }
  },

  submitClassroom() {
    this.setState({ loading: true, });
    const that = this;
    $.post('/teachers/classrooms/', { classroom: this.state.classroom, })
        .success((data) => {
          let nextPage;
          if (that.props.closeModal) {
              // only used if it is rendered within a modal
            that.props.closeModal('because class added');
          } else if (that.props.hasActivities === false) {
            window.location.assign('/teachers/classrooms/assign_activities');
          } else if (data.toInviteStudents) {
            window.location.assign('/teachers/classrooms/invite_students');
          } else {
            window.location.assign('/teachers/classrooms/scorebook');
          }
        })
        .error((data) => {
          that.setState({ loading: false, });
          const errors = JSON.parse(data.responseText).errors;
          const errorMessage = errors.join('\n');
          that.setState({ errors: errorMessage, });
        });
  },

  syncOrModal() {
    if (this.props.user.signed_up_with_google) {
        // they are already a google user, so we just need to use the callback
      this.syncClassrooms();
    } else {
        // they are not a google user, so we will show them the modal where they
        // can become one
      this.setState({ showModal: true, });
    }
  },

  syncClassrooms() {
    window.location = '/teachers/classrooms/google_sync';
  },

  hideModal() {
    this.setState({ showModal: false, });
  },

  createButton() {
    return this.state.loading
    ? <button className="button-grey submit-button"><ButtonLoadingIndicator/></button>
    : <button className="button-green submit-button" onClick={this.clickCreateAClass}>Create a Class</button>
  },

  formatTitle() {
    const classroom = this.state.classroom;
    if (classroom.grade) {
      return classroom.grade == 'University' || classroom.grade == 'Other' ? classroom.grade : `${NumberSuffix(classroom.grade)} Grade`;
    }
    return 'Select Grade';
  },

  render() {
    const classroom = this.state.classroom;
    return (
      <div className="create-a-class-page">
        <div className="create-a-class">
          <div>
            <div id="new-classroom">
              <h1>Create Your Class</h1>
              <p>After you create a class, you can create studentsâ€™ accounts or have your students sign up with a class code.</p>
              <div className="input-row">
                <label htmlFor="class-name">Class Name:</label>
                <input type="text" id="class-name" placeholder="e.g. 6th Period ELA" value={classroom.name} onChange={this.handleChange} />
              </div>
              <div className="input-row">
                <label htmlFor="">Grade:</label>
                <DropdownButton className={classroom.grade ? 'has-grade' : null} bsStyle="default" id="select-grade" title={this.formatTitle()} onSelect={this.handleSelect}>
                  {this.grades()}
                </DropdownButton>
              </div>
              <div className="input-row">
                <label htmlFor="classroom_code">Class Code:</label>
                <input className="inactive class-code text-center" disabled="true" type="text" value={classroom.code} name="classroom[code]" id="classroom_code" />
              </div>
              <div id="regenerate-class-code" onClick={this.getClassCode}><span><i className="fa fa-refresh" />Regenerate Class Code</span></div>
              {this.createButton()}
              <h4 className="errors">{this.state.errors}</h4>
            </div>
            <div id="new-google-classroom">
              <h1>Google Classroom User? <img src="/images/google_classroom_icon.png" data-pin-nopin="true" /></h1>
              <p>If you have an account with Google Classroom, you can import all your classes and students to Quill.</p>
              <button className="button-green" onClick={this.syncOrModal}>Import From Google Classroom</button>
              <GoogleClassroomModal syncClassrooms={this.syncClassrooms} user={this.props.user} show={this.state.showModal} hideModal={this.hideModal} />
            </div>
          </div>
        </div>
      </div>);
  },
});
