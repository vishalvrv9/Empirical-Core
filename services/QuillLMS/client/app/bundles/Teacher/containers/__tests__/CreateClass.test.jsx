import React from 'react';
import { shallow } from 'enzyme';

import CreateClass from '../CreateClass.jsx';

import $ from 'jquery'
import DropdownButton from 'react-bootstrap/lib/DropdownButton';
import MenuItem from 'react-bootstrap/lib/MenuItem';
import ButtonLoadingIndicator from '../../components/shared/button_loading_indicator.jsx'
import GoogleClassroomModal from '../../components/dashboard/google_classroom_modal'

jest.mock('jquery', () => {
  return {
    ajax: jest.fn(),
    post: jest.fn().mockReturnValue({
      success: jest.fn().mockReturnValue({
        error: jest.fn()
      })
    })
  }
});

describe('CreateClass container', () => {

  it('should display a ButtonLoadingIndicator component if loading', () => {
    const wrapper = shallow(<CreateClass />);
    wrapper.setState({loading: true});
    expect(wrapper.find(ButtonLoadingIndicator).exists()).toBe(true);
  });

  it('should render a blank create class form if no class exists in state', () => {
    expect(shallow(<CreateClass />)).toMatchSnapshot();
  })

  describe('with class in state', () => {
    const wrapper = shallow(
      <CreateClass
        user={{foo: 'bar'}}
      />
    );
    wrapper.setState({
      classroom: {
        name: 'English 101',
        grade: 7,
        code: 'smelly-fish'
      },
      loading: false
    });

    describe('classroom name text input', () => {
      it('should render with classroom name as value', () => {
        expect(wrapper.find('#class-name').props().value).toBe('English 101');
      });

      it('onChange should update classroom name', () => {
        wrapper.find('#class-name').simulate('change', {currentTarget: {value: 'English 404'}});
        expect(wrapper.state().classroom.name).toBe('English 404');
      });
    });

    describe('DropdownButton component', () => {
      it('should have correctly formatted title prop', () => {
        expect(wrapper.find(DropdownButton).props().title).toBe('7th Grade');
        wrapper.setState({ classroom: { grade: 1 }});
        expect(wrapper.find(DropdownButton).props().title).toBe('1st Grade');
        wrapper.setState({ classroom: { grade: 2 }});
        expect(wrapper.find(DropdownButton).props().title).toBe('2nd Grade');
        wrapper.setState({ classroom: { grade: 3 }});
        expect(wrapper.find(DropdownButton).props().title).toBe('3rd Grade');
        wrapper.setState({ classroom: { grade: 'University' }});
        expect(wrapper.find(DropdownButton).props().title).toBe('University');
        wrapper.setState({ classroom: { grade: 'Other' }});
        expect(wrapper.find(DropdownButton).props().title).toBe('Other');
        wrapper.setState({
          classroom: {
            name: 'English 101',
            grade: 7,
            code: 'smelly-fish'
          }
        });
      });

      it('should render MenuItems for grades K-12, University, and Other', () => {
        expect(wrapper.find(MenuItem)).toHaveLength(14);
      });

      it('onSelect event should change grade in state', () => {
        wrapper.find(DropdownButton).simulate('select', 6);
        expect(wrapper.state().classroom.grade).toBe(6);
      });
    });

    it('classroom code should render', () => {
      expect(wrapper.find('#classroom_code').props().value).toBe('smelly-fish');
    });

    it('regeneration button onClick should retrieve new code and call setClassCode on success', () => {
      wrapper.find('#regenerate-class-code').simulate('click');
      expect($.ajax).toHaveBeenCalled();
      expect($.ajax.mock.calls[0][0].url).toBe('/teachers/classrooms/regenerate_code');
      expect($.ajax.mock.calls[0][0].success).toBe(wrapper.instance().setClassCode);
    });

    describe('create a class button onClick', () => {
      it('should call submitClassroom if there is a name and grade', () => {
        wrapper.instance().submitClassroom = jest.fn();
        wrapper.find('.submit-button').simulate('click');
        expect(wrapper.instance().submitClassroom).toHaveBeenCalled;
      });

      it('should set errors if name is missing', () => {
        const wrapperWithoutName = shallow(
          <CreateClass />
        );
        wrapperWithoutName.setState({
          classroom: {
            grade: 7,
            code: 'smelly-fish'
          },
          loading: false
        });
        wrapperWithoutName.find('.submit-button').simulate('click');
        expect(wrapperWithoutName.state().errors).toBe('You must choose your classroom\'s name before continuing.');
      });

      it('should set errors if grade is missing', () => {
        const wrapperWithoutGrade = shallow(
          <CreateClass />
        );
        wrapperWithoutGrade.setState({
          classroom: {
            name: 'English 101',
            code: 'smelly-fish'
          },
          loading: false
        });
        wrapperWithoutGrade.find('.submit-button').simulate('click');
        expect(wrapperWithoutGrade.state().errors).toBe('You must choose your classroom\'s grade before continuing.');
      });
    });

    it('errors should render', () => {
      expect(wrapper.find('.errors').text()).toBe('');
      wrapper.setState({errors: 'Everything is broken!'});
      expect(wrapper.find('.errors').text()).toBe('Everything is broken!');
    })

    describe('import from Google button onClick', () => {
      const wrapperSignedUpWithGoogle = shallow(
        <CreateClass
          user={{signed_up_with_google: true}}
        />
      );
      wrapperSignedUpWithGoogle.instance().syncClassrooms = jest.fn();
      const wrapperNotSignedUpWithGoogle = shallow(
        <CreateClass
          user={{signed_up_with_google: false}}
        />
      );
      [wrapperSignedUpWithGoogle, wrapperNotSignedUpWithGoogle].forEach((w) => {
        w.setState({
          classroom: {
            name: 'English 101',
            grade: 7,
            code: 'smelly-fish'
          },
          loading: false
        });
        w.find('#new-google-classroom button').simulate('click');
      });

      it('should call syncClassrooms if user is signed up with google', () => {
        expect(wrapperSignedUpWithGoogle.instance().syncClassrooms).toHaveBeenCalled();
      });

      it('should show modal if user is not signed up with google', () => {
        expect(wrapperNotSignedUpWithGoogle.state().showModal).toBe(true);
      });
    });

    describe('GoogleClassroomModal component', () => {
      it('should have syncClassrooms prop', () => {
        //TODO: determine how best to test syncClassrooms actually redirects w/ window.location
        expect(wrapper.find(GoogleClassroomModal).props().syncClassrooms).toBe(wrapper.instance().syncClassrooms);
      });

      it('should have user prop', () => {
        expect(wrapper.find(GoogleClassroomModal).props().user.foo).toBe('bar');
      });

      it('should have show prop based on state.showModal', () => {
        wrapper.setState({showModal: false});
        expect(wrapper.find(GoogleClassroomModal).props().show).toBe(false);
        wrapper.setState({showModal: true});
        expect(wrapper.find(GoogleClassroomModal).props().show).toBe(true);

      });

      it('should have hideModal prop that hides modal', () => {
        wrapper.setState({showModal: true});
        wrapper.find(GoogleClassroomModal).props().hideModal();
        expect(wrapper.state().showModal).toBe(false);
      });
    });

    describe('submitClassroom function', () => {
      //TODO: add tests for contents of success and error functions?
      const wrapper = shallow(<CreateClass />);
      it('should send post request with classroom data', () => {
        wrapper.setState({
          classroom: {
            name: 'English 202',
            grade: 9,
            code: 'cute-puppy'
          }
        });
        wrapper.instance().submitClassroom();
        expect($.post).toHaveBeenCalled();
        expect($.post.mock.calls[0][0]).toBe('/teachers/classrooms/');
        expect($.post.mock.calls[0][1].classroom.name).toBe('English 202');
        expect($.post.mock.calls[0][1].classroom.grade).toBe(9);
        expect($.post.mock.calls[0][1].classroom.code).toBe('cute-puppy');
      });
    });

  });

});
