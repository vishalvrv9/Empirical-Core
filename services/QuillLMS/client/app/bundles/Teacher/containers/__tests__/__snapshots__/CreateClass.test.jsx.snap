// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateClass container should render a blank create class form if no class exists in state 1`] = `
<div
  className="create-a-class-page"
>
  <div
    className="create-a-class"
  >
    <div>
      <div
        id="new-classroom"
      >
        <h1>
          Create Your Class
        </h1>
        <p>
          After you create a class, you can create studentsâ€™ accounts or have your students sign up with a class code.
        </p>
        <div
          className="input-row"
        >
          <label
            htmlFor="class-name"
          >
            Class Name:
          </label>
          <input
            id="class-name"
            onChange={[Function]}
            placeholder="e.g. 6th Period ELA"
            type="text"
            value=""
          />
        </div>
        <div
          className="input-row"
        >
          <label
            htmlFor=""
          >
            Grade:
          </label>
          <DropdownButton
            bsStyle="default"
            className={null}
            id="select-grade"
            onSelect={[Function]}
            title="Select Grade"
          >
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={1}
              header={false}
            >
              1st
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={2}
              header={false}
            >
              2nd
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={3}
              header={false}
            >
              3rd
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={4}
              header={false}
            >
              4th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={5}
              header={false}
            >
              5th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={6}
              header={false}
            >
              6th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={7}
              header={false}
            >
              7th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={8}
              header={false}
            >
              8th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={9}
              header={false}
            >
              9th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={10}
              header={false}
            >
              10th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={11}
              header={false}
            >
              11th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey={12}
              header={false}
            >
              12th
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="University"
              header={false}
            >
              University
            </MenuItem>
            <MenuItem
              bsClass="dropdown"
              disabled={false}
              divider={false}
              eventKey="Other"
              header={false}
            >
              Other
            </MenuItem>
          </DropdownButton>
        </div>
        <div
          className="input-row"
        >
          <label
            htmlFor="classroom_code"
          >
            Class Code:
          </label>
          <input
            className="inactive class-code text-center"
            disabled="true"
            id="classroom_code"
            name="classroom[code]"
            type="text"
            value={null}
          />
        </div>
        <div
          id="regenerate-class-code"
          onClick={[Function]}
        >
          <span>
            <i
              className="fa fa-refresh"
            />
            Regenerate Class Code
          </span>
        </div>
        <button
          className="button-green submit-button"
          onClick={[Function]}
        >
          Create a Class
        </button>
        <h4
          className="errors"
        />
      </div>
      <div
        id="new-google-classroom"
      >
        <h1>
          Google Classroom User? 
          <img
            data-pin-nopin="true"
            src="/images/google_classroom_icon.png"
          />
        </h1>
        <p>
          If you have an account with Google Classroom, you can import all your classes and students to Quill.
        </p>
        <button
          className="button-green"
          onClick={[Function]}
        >
          Import From Google Classroom
        </button>
        <google_classroom_modal
          hideModal={[Function]}
          show={false}
          syncClassrooms={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
